File: apps/ios/Barns/BarnsApp/Presentation/Auth/AuthViewModel.swift
import Combine

@MainActor
final class AuthViewModel: ObservableObject {
    enum State: Equatable {
        case idle
        case loading
        case authenticated(User)
        case failed(String)
    }

    @Published private(set) var state: State = .idle

    private let loginAsGuestUseCase: LoginAsGuestUseCase
    private let getCurrentUserUseCase: GetCurrentUserUseCase

    init(
        loginAsGuestUseCase: LoginAsGuestUseCase,
        getCurrentUserUseCase: GetCurrentUserUseCase
    ) {
        self.loginAsGuestUseCase = loginAsGuestUseCase
        self.getCurrentUserUseCase = getCurrentUserUseCase
    }

    var isAuthenticated: Bool {
        if case .authenticated = state { return true }
        return false
    }

    func loginAsGuest() async {
        state = .loading
        do {
            let user = try await loginAsGuestUseCase.execute()
            state = .authenticated(user)
        } catch {
            state = .failed("Sign-in failed. Please try again.")
        }
    }
}

File: apps/android/app/src/main/java/com/barns/app/presentation/auth/AuthViewModel.kt
package com.barns.app.presentation.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.barns.app.domain.model.User
import com.barns.app.domain.usecase.auth.GetCurrentUserUseCase
import com.barns.app.domain.usecase.auth.LoginAsGuestUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class AuthViewModel(
    private val loginAsGuestUseCase: LoginAsGuestUseCase,
    private val getCurrentUserUseCase: GetCurrentUserUseCase,
) : ViewModel() {
    sealed interface State {
        object Idle : State
        object Loading : State
        data class Authenticated(val user: User) : State
        data class Failed(val message: String) : State
    }

    private val _state = MutableStateFlow<State>(State.Idle)
    val state: StateFlow<State> = _state.asStateFlow()

    val isAuthenticated: Boolean
        get() = _state.value is State.Authenticated

    fun loginAsGuest() {
        _state.value = State.Loading
        viewModelScope.launch {
            runCatching { loginAsGuestUseCase.execute() }
                .onSuccess { user -> _state.value = State.Authenticated(user) }
                .onFailure { _state.value = State.Failed("Sign-in failed. Please try again.") }
        }
    }

    fun loadCurrentUser() {
        viewModelScope.launch {
            runCatching { getCurrentUserUseCase.execute() }
                .onSuccess { user -> if (user != null) _state.value = State.Authenticated(user) }
                .onFailure { _state.value = State.Failed("Sign-in failed. Please try again.") }
        }
    }
}
